beater/middleware/rate_limit_middleware.go

```diff
@@ -0,0 +1,38 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package middleware
+
+import (
+	"github.com/elastic/apm-server/beater/api/ratelimit"
+	"github.com/elastic/apm-server/beater/config"
+	"github.com/elastic/apm-server/beater/request"
+)
+
+const burstMultiplier = 3
+
+// RateLimitMiddleware adds a rate limiter per IP to the request context
+func RateLimitMiddleware(cfg *config.EventRate) Middleware {
```

Comment:
Providing a `RateLimiterMiddleware` I would expect the middleware to take care of the rate limiting already and not only of setting a rate limit manager. The current rate Intake API doesn't really allow that, as the rate limiting occurs depending on the amount of events. 

For something like e.g. ACM for RUM we will probably be able to apply an easier rate limiting strategy as it doesn't need to throttle processing of events, but can simply be an allow or deny strategy (similar to the killswitch middleware but asking a rate limiter). I'd assume this will be doable in a rate limit middleware without carrying logic over to the handler. 

I assume the refactoring is based on the rate limiter needs for RUM ACM. As pointed out I don't think that this will require the same rate limiting handling as for Intake API, so I don't see a lot of value in this concrete middleware. If you still want to move forward with it, then please rename to something that makes it clear that this only _sets_ a rate limiter but doesn't apply it.

Reply:
It will still require a rate limiter per IP, and like this we can reuse as much logic as possible. But since everything else is a middleware also, it also makes sense to treat rate limiting in the same way. It is not so different eg. than the logging middleware, which puts some stuff in the context and it gets used later on

Reply:
The purpose of a logging middleware is to log something and not to decline a request, so I don't think the comparison is appropriate. 

I do agree that when having an allow/deny policy for a rate limiter we should definitly add this as a middleware. That would be a different one though (as pointed out above). 
As said before, not against it, just don't see a lot of value in it and definitely would rename to something more clear.

Reply:
> just don't see a lot of value in it

The value is code reuse, and consistency: extract functionality that can extend the capabilities of any handler, without knowledge of the handler internals. 
Would you prefer to hold onto this until see how is applied to the new endpoint?

> rename to something more clear

"RateLimiterManagerMiddleware"? :thinking:

Reply:
`SetRateLimiterMiddleware`?

Reply:
Having a verb there reads weird to me. The comparison with the logging middleware was because it creates a logger, sets it in the context, and then it is read/used from handlers (_and_ also from the middleware, yes). Yet it is not called `SetLoggerMiddleware`.

The reason why names are confusing to me is the "manager" bit. The `RateLimitManager` is effectively a rate limiter per IP, I don't understand why `"per IP" == "Manager", the fact that there is a cache + logic that _manages_ something  is an unimportant detail, the important detail is the "per IP" part.
So why not call the whole thing `IPRateLimiter`, and then this `IPRateLimitMiddleware`? 

Ideally it should also be dealing with the "per IP" part, moving this logic https://github.com/elastic/apm-server/blob/0ccad2049ec81008528074d67fe112aeb2e5d6aa/beater/api/intake/handler.go#L57 (and the `rateLimit` implementation) to the middleware, and setting in the context the specific `rate.Limiter` for that request. 

It looks like your confusion comes from the fact that the middleware doesn't do much, would that help?
Again, I didn't want to make this bigger than strictly needed but I am happy to do that just here or in another PR.

Reply:
The log middleware is setting the logger, but it also takes care of all the logging during the request handling. The only reason it also sets the logger in the context is to ensure the edge case of a response not being able to be written is covered. I still do not agree that this comparison is a good one. 
You can compare it with the [SetAuthorizationMiddleware](https://github.com/elastic/apm-server/blob/master/beater/middleware/authorization_middleware.go#L50). Two authorization middlewares are available, one for only setting flags and one for rejecting the whole request.

Let's focus on the middleware naming first: My point is, this middleware doesn't do anything other than setting a value. From the name of the middleware I expect the middleware to also apply the rate limiting, which it doesn't. This needs to be clarified.

Regarding your second point:
The `RateLimiterManager` currently is an interface with one method `RateLimiter`. It is not concerned _at all_ with what the key for rate limiting is. And it shouldn't. The IP part is handled by the API [handler](https://github.com/elastic/apm-server/blob/master/beater/api/intake/handler.go#L155). Now when reusing the whole rate limiting logic, it might make sense to build some IPRateLimiter logic that composes the existing rate limiter logic to one logic and then use this everywhere. But it should not be mixed into the current rate limiter manager. Let's keep concerns separated.

Reply:
> Again, I didn't want to make this bigger than strictly needed but I am happy to do that just here or in another PR.

+1 on not making this bigger as necessary here. I am not saying you should do more refactoring, only advocating for a different name.

Reply:
I can put another example: all that `RequestTimeMiddleware` does is set `time.Now()` in the request context, yet it is not called `SetRequestTimeMiddleware`.

My point about names is that the should convey _what_ is done, not _how_ something is done. Setting a variable is an implementation detail (_how_), and someone new to the codebase wouldn't find `SetXXXMiddleware` any more enlightening.

I generally don't complaint about names unless they lead me in the wrong direction, which is the case for this suggestion: `SetXXX` is a widespread naming convention for setters, which is not what the middleware is. Functions called `SetXXX` that return something are always confusing to me...

Similar thing goes for `RateLimiterManager`, that "Manager" suffix doesn't help to understand _what_ is happening inside, it is not meaningful. Interface should be named `RateLimiter`, define the method `RateLimit`, and concrete implementation should be `IPRateLimiter` (see eg. in https://talks.golang.org/2014/names.slide#13).

Having several not-so-clear names in the same workflow (like "RateLimiterManager" and then "SetRateLimitMiddleware") undermines a bit the understanding of it.